doc: add information about modules cache behavior (#26971)
nodejs/node: doc/api/modules.md

this does not apply to [native addons][], for which reloading will result in an
error.

Adding or replacing entries is also possible. This cache is checked before

> native modules and if such a name is added to the cache, no require call is

Reviewer: nit: `such` here can be ambiguous, depending on the reader.

```suggestion
native modules and if a name matching a native module is added to the cache,
no require call is
```

require.extensions['.sjs'] = require.extensions['.js'];
```

**Deprecated.** In the past, this list has been used to load non-JavaScript
modules into Node.js by compiling them on-demand. However, in practice, there
are much better ways to do this, such as loading modules via some other Node.js
program, or compiling them to JavaScript ahead of time.

> Avoid using `require.extensions`. Use could cause subtle bugs and resolving the

Reviewer: nit:
```suggestion
Avoid using `require.extensions`. It could cause subtle bugs and resolving the
```

> Avoid using `require.extensions`. Use could cause subtle bugs and resolving the

Reviewer: nit: Reword for simplicity.
```suggestion
Avoid using `require.extensions`. It can cause subtle bugs and resolving
```

Author: Thanks, I personally prefer `Use` in this specific case but I incorporated your other suggestions.

> extensions is becoming slower the more extensions you add.

Reviewer: continued from the above nit:
```suggestion
extensions gets slower with each registered extension.
```
